File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.js
/**
 * // Definition for a Node.
 * function Node(val, next, random) {
 *    this.val = val;
 *    this.next = next;
 *    this.random = random;
 * };
 */

/**
 * @param {Node} head
 * @return {Node}
 */
var copyRandomList = function(head) {

  if (head === null) return null;

  const newHead = new Node(head.val, null, null);
  newHead.ref = head;

  // const linkedListNodes = [newHead];

  let curr = head.next;
  let prev = newHead;

  while (curr) {
    const newNode = new Node(curr.val, null, null);
    newNode.ref = curr;

    prev.next = newNode;

    prev = newNode;

    // linkedListNodes.push(newNode);

    curr = curr.next;

  }

  curr = newHead;
  readCurr = head;

  while (readCurr) {
    const randomNode = readCurr.random;

    if (randomNode === null) curr.random = null;
    else {
      let iCurr = newHead;
      while(iCurr) {
        if (iCurr.ref === randomNode) {
          curr.random = iCurr;
          break;
        }
        iCurr = iCurr.next;
      }
    }

    curr = curr.next;
    readCurr = readCurr.next;
  }

  return newHead;
};

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""


class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':

        if head is None:
            return None

        new_head = Node(head.val)
        prev_node = new_head

        curr = head.next

        while curr:
            new_node = Node(curr.val)
            prev_node.next = new_node
            prev_node = new_node
            curr = curr.next

        curr = head
        curr_new = new_head

        while curr:
            if curr.random is None:
                curr = curr.next
                curr_new = curr_new.next
                continue

            random_node = curr.random

            scan_curr = head
            new_scan_curr = new_head

            while scan_curr:
                if scan_curr == random_node:
                    curr_new.random = new_scan_curr
                    break

                scan_curr = scan_curr.next
                new_scan_curr = new_scan_curr.next

            curr = curr.next
            curr_new = curr_new.next

        return new_head
